src/db/models/device.rs

```diff
@@ -777,68 +795,118 @@ impl Device<Id> {
         Ok((network_info, configs))
     }
 
-    // Assign IP to the device in a given network
+    /// Assign the next available IP address in each subnet of the network to this device.
+    ///
+    /// For every CIDR block in `network.address`, this function:
+    /// 1. Iterates through the block's IPs in order.
+    /// 2. Skips any IP that:
+    ///    - Fails the `can_assign_ips` validation (out of range, reserved, or already in use by another device), or  
+    ///    - Appears in the optional `reserved_ips`.
+    /// 3. Selects the first remaining IP and records it.
+    ///
+    /// If any subnet has no valid, unassigned IP, the method returns `ModelError::CannotCreate`.
+    ///
+    /// # Parameters
+    ///
+    /// - `transaction`: Active PostgreSQL connection to check and insert assignments.  
+    /// - `network`: The `WireguardNetwork<Id>` whose subnets will be assigned.  
+    /// - `reserved_ips`: Optional slice of IPs that must not be assigned, even if otherwise free.
+    ///
+    /// # Returns
+    ///
+    /// - `Ok(WireguardNetworkDevice)`: A new relation linking this device to its assigned IPs across all subnets.  
+    /// - `Err(ModelError::CannotCreate)`: If any subnet lacks an available IP.
     pub(crate) async fn assign_next_network_ip(
         &self,
         transaction: &mut PgConnection,
         network: &WireguardNetwork<Id>,
         reserved_ips: Option<&[IpAddr]>,
     ) -> Result<WireguardNetworkDevice, ModelError> {
-        if let Some(address) = network.address.first() {
-            let net_ip = address.ip();
-            let net_network = address.network();
-            let net_broadcast = address.broadcast();
-            for ip in address {
-                if ip == net_ip || ip == net_network || ip == net_broadcast {
-                    continue;
-                }
-                if let Some(reserved_ips) = reserved_ips {
-                    if reserved_ips.contains(&ip) {
-                        continue;
-                    }
-                }
+        debug!(
+            "Assiging IP addresses for device: {} in network {}",
+            self.name, network.name
+        );
+        let mut ips = Vec::new();
+        let reserved = reserved_ips.unwrap_or(&[]);
```

Comment:
Default is an empty slice:

```suggestion
        let reserved = reserved_ips.unwrap_or_default();
```